Create staging folder for repo k8s.io/api, move external API used by client-go to k8s.io/api

caesarxuchao opened the pull request:
Comments that will be addressed in followup PR:
- [ ] fix import lines like `clientv1 "k8s.io/api/core/v1"`
- [ ] fix import line `"k8s.io/api/core/v1"` that belongs to another import section
- [ ] Export functions instead of making SchemeBuilder public. ([Discussion](https://github.com/kubernetes/kubernetes/commit/a6c97715ed7e24581dd9a02a256d57e56b70ac44#commitcomment-22146536) in #45463 and #45910)
- [ ] generate clientset to staging/client-go, get rid of the copy.


Fix https://github.com/kubernetes/kubernetes/issues/44065

**What's moved:**
types.go, constants for annotation keys, methods of the types, validation, defaults, generated swagger doc

**What tests are passed:**
Unit tests are passed

**Next steps:**
1. fix all the code generators (@mhbauer are you interested in helping fixing them?)
- [x] deepcopy-gen: i fixed it locally.
- [x] conversion-gen: fixed locally
- [ ] client-gen
...

2. open PRs to move other api groups
3. disable the publish robot
4. merge all PRs
5. create the k8s.io/api repo for real, and starts
6. re-enable the publish robot

```release-note
The schema of the API that are served by the kube-apiserver, together with a small amount of generated code, are moved to k8s.io/api (https://github.com/kubernetes/api).
```

Commit:
manually fix unit tests in pkg/api/v1

more manually fix pkg/api/v1 unit tests

change ImportPrefix

manually fix pkg/api unit tests

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44784)
<!-- Reviewable:end -->

Commit:
split v1/register.go to regsiter.go and builder.go
move api/v1 to k8s.io/api/v1
duplicate some global variables/functions in pkg/api/v1/builder.go, add todo to remove these

Commit:
let conversion-gen to choose localSchemeBuiler

Commit:
vendorless path for informer-gen

Commit:
some copy.sh changes

caesarxuchao commented:
make all && make test passed locally.

Commit:
manually fix protogen

caesarxuchao commented:
This PR is working locally, just needs a rebase.

I plan to get this merged after code freeze, to avoid disrupting other developers in 1.7. I'll bring this up in the sig-apimachinery meeting and the community meeting.

seh commented on cmd/kubeadm/app/phases/apiconfig/clusterroles.go:
```diff
@@ -19,10 +19,11 @@ package apiconfig
 
+	"k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1beta1"
+rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1beta1"
```
Is that GitHub displaying the tabs incorrectly, or is "rbachelper" really at the start of the line?

caesarxuchao replied:
I'll run `find *.go | xargs gofmt -w` before really submitting this.

caesarxuchao replied:
@seh i couldn't find your other comment about exporting a function instead of SchemeBuilder in register.go, so i reply here. That will require additional changes to the code generators (deepcopy, conversion, etc), some (deepcopy-gen) are not in the kubernetes repo and will thus require multiple PRs, so i decided not to do that refactor in this PR.

seh replied:
Here's [a link to the comment](https://github.com/kubernetes/kubernetes/pull/44784#discussion_r121678895) against file _staging/src/k8s.io/api/apps/v1beta1/register.go_, which GitHub hides because the commit is outdated.

Commit:
Don't revert, necessary change to make helpers to include k8s.io/api

    Don't revert, change boundingdirs in Makefile for deepcopy-gen

manually fix pkg/client/clientset_generated/clientset/typed/core/v1/pod_expansion.go because external policy types are moved now

Commit:
unify register.go formats:
networking/v1/register.go
staging/src/k8s.io/kube-apiextensions-server/pkg/apis/apiextensions/v1alpha1/register.go
staging/src/k8s.io/metrics/pkg/apis/custom_metrics/v1alpha1/register.go
staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1 register.go

Commit:
revert before merge? remove a dependency from pkg/apis/componentconfig to clientset, probably we should fix it later. i removed the dependency to test if pkg/apis compile

Commit:
manually fix unit tests

caesarxuchao commented:
Ok, it's a rebase hell. I'll do another one over the weekend.

Commit:
hack/lib/init.sh util.sh update-codegen.sh

Commit:
manually remove pkg/client/listers/policy/v1alpha1

caesarxuchao commented:
@k8s-bot pull-kubernetes-unit test this

Commit:
change hack/update-codecgen.sh

Commit:
manually fix hack/verify-staging-imports.sh

manually fix hack/godep-save.sh

manually fix .linted_packages

squash, manually fix lint

manually fix hack/update-staging-godeps.sh

manually fix update-staging-godeps.sh; let it update godep of client-go

Commit:
manually fix kubectl openapi unit test

Commit:
manually fix openapi-gen

caesarxuchao commented:
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

caesarxuchao commented on pkg/kubectl/cmd/util/openapi/openapi_test.go:
```diff
@@ -34,9 +34,10 @@ var _ = Describe("Reading apps/v1beta1/Deployment from openAPIData", func() {
 	})
 
-	deploymentName := "io.k8s.kubernetes.pkg.apis.apps.v1beta1.Deployment"
+	deploymentName := "io.k8s.api.apps.v1beta1.Deployment"
```
@mbohlool With the external API moved to k8s.io/api, the names in openapi change. Is there any backward compatibility concern?

lavalamp replied:
Open an issue about this, we can fix it after the fact if it is a problem.

caesarxuchao replied:
https://github.com/kubernetes/kubernetes/issues/47372

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go:
```diff
@@ -139,7 +139,11 @@ func friendlyName(name string) string {
-	return fmt.Sprintf("%s.%s", t.PkgPath(), t.Name())
+	path := t.PkgPath()
+	if strings.Contains(path, "/vendor/") {
+		path = path[strings.Index(path, "/vendor/")+len("/vendor/"):]
```
This makes the `typeName` the same as the other [helper](https://github.com/kubernetes/kubernetes/blob/0cff83931739cb7dbf344b502540d7e342eeaf04/staging/src/k8s.io/apiserver/pkg/server/openapi/openapi.go#L118-L121) used in another package.

I can leave a todo to move the util to a common place. cc @mbohlool

caesarxuchao commented:
Hmm, need to fix the Godep/ of client-go.

caesarxuchao commented:
@k8s-bot pull-kubernetes-unit test this

caesarxuchao commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

lavalamp commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -30,7 +30,7 @@ import (
 	"k8s.io/apiserver/pkg/server/healthz"
 	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
-	clientv1 "k8s.io/client-go/pkg/api/v1"
+	clientv1 "k8s.io/api/core/v1"
```
clientv1 doesn't seem like a good name any more but we can fix that later.

lavalamp commented on pkg/apis/apps/v1beta1/register.go:
```diff
@@ -34,32 +34,14 @@ func Resource(resource string) schema.GroupResource {
 	// generated functions takes place in the generated files. The separation
 	// makes the code compile even when the generated files are missing.
-	localSchemeBuilder.Register(addKnownTypes, addDefaultingFuncs, addConversionFuncs)
+	localSchemeBuilder.Register(appsv1beta1.AddKnownTypes, addDefaultingFuncs, addConversionFuncs)
```
Don't you have the external types already adding AddKnownTypes? Won't this double-add it?

caesarxuchao replied:
Will fix.

lavalamp commented on staging/src/k8s.io/api/apps/v1beta1/register.go:
```diff
@@ -0,0 +1,65 @@
+var (
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
+	SchemeBuilder      runtime.SchemeBuilder
```
You can change this to `SchemeBuilder = runtime.NewSchemeBuilder(AddKnownTypes)` and delete the init function.

caesarxuchao replied:
I'll see if it's easy to change next time I rebase.

seh replied:
Related:
- [Discussion](https://github.com/kubernetes/kubernetes/commit/a6c97715ed7e24581dd9a02a256d57e56b70ac44#commitcomment-22146536) in #45463
- #45910

caesarxuchao replied:
Ack. I'll either do it in the next rebase or in a follow up.

lavalamp commented on staging/src/k8s.io/api/apps/v1beta1/register.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+// Adds the list of known types to api.Scheme.
+func AddKnownTypes(scheme *runtime.Scheme) error {
```
Do these really need to be public?

lavalamp commented on staging/src/k8s.io/api/apps/v1beta1/register.go:
```diff
@@ -0,0 +1,65 @@
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
+	SchemeBuilder      runtime.SchemeBuilder
+	localSchemeBuilder = &SchemeBuilder
```
What is the benefit of the localSchemeBuilder alias?

caesarxuchao replied:
I guess you saw why I did this after you reviewed the later commits.

lavalamp commented:
I left comments, but github is hiding some of them because I made them on individual commits...

caesarxuchao commented:
Still need to fix staging/copy.sh so that it doesn't remove the Godeps, the PR otherwise pass all tests.

lavalamp commented:
Yeah other than the SchemeBuilder thing this looks good--consider this an lgtm once that is fixed.

Commit:
revert!! temporary change to hack/update-all.sh

Commit:
revert hack/update-all to its original form

Commit:
run update-staging-godeps.sh; the script asks user to manually commit changes for each repo

caesarxuchao commented:
All tests passed.

caesarxuchao commented:
/test pull-kubernetes-cross

caesarxuchao commented:
All tests passed again. Will rebase tomorrow, after we disable the submit-queue.

Commit:
run pkg/api/v1/rewrite....sh, pkg/api/v1 (not including subdir) compile

Commit:
run root-rewrite-v1-..., compile

Commit:
pkg/apis/move-external-types-for-apis.sh, k8s.io/api compiles

Commit:
run pkg/apis/make-origin-dir-compile, pkg/apis compile

Commit:
run ./root-rewrite-all-other-apis.sh, then run make all, pkg/... compiles

Commit:
run fix-casttype.sh

Commit:
run hack/update-codegen.sh

Commit:
run ./remove-original-proto.sh to remove the old proto

Commit:
run root-rewrite-import-client-go-api-types

Commit:
run hack/update-staging-client-go, somehow we copied listers/<authn,authz,imagepolicy>

Commit:
make all works. generated harmless covnersion/deepcoy chagnes

caesarxuchao commented:
The SchemeBuilder issue is fixed.

Commit:
run hack/update-all

caesarxuchao commented:
Rebased.

caesarxuchao commented:
Squashed a little. I don't squash it further because i might need to do another rebase, or revert some changes.

caesarxuchao commented:
@lavalamp i'm adding the lgtm label based on this https://github.com/kubernetes/kubernetes/pull/44784#issuecomment-307949929. We'll merge the PR today.

lavalamp commented:
/approve

caesarxuchao commented:
Submit queue is suspended. Going to manually merge this PR, then resume the submit-queue, with code freeze lifted.

mikedanese commented:
Can we move this into 1.7 branch? This is going to cause a TON of merge conflicts for cherrypicks if we don't put this in 1.7.
